frontend/controller/actions/group.js

```diff
@@ -321,6 +322,36 @@ export default (sbp('sbp/selectors/register', {
       throw new GIErrorUIRuntimeError(L('Failed to join chat channel.'))
     }
   },
+  'gi.actions/group/leaveChatRoom': async function (params: GIActionParams) {
+    try {
+      const rootState = sbp('state/vuex/state')
+      const { username, chatRoomID } = params.data
+
+      await sbp('gi.actions/chatroom/leave', {
+        contractID: chatRoomID,
+        data: {
+          username,
+          operator: params.options?.leavingGroup ? username : rootState.loggedIn.username
+        },
+        hooks: {
+          prepublish: params.hooks?.prepublish,
+          postpublish: null
+        }
+      })
+
+      await sbp('chelonia/out/actionEncrypted', {
+        ...omit(params, ['options', 'action', 'hooks']),
+        action: 'gi.contracts/group/leaveChatRoom',
+        hooks: {
+          prepublish: null,
+          postpublish: params.hooks?.postpublish
+        }
+      })
```

Comment:
Could you please re-send this PR, building on top of `e2e-protocol` instead of `master`?

The way these inner messages are done in `e2e-protocol` is different, and we would need to address merge conflicts if this PR is merged.

In `e2e-protocol`, instead of calling `'chelonia/out/actionEncrypted'` directly yourself, you call a function called `sendMessage`. Here's an example of how `'gi.actions/group/renameChatRoom'` has been re-implemented:

```js
  ...encryptedAction('gi.actions/group/renameChatRoom', L('Failed to rename chat channel.'), async function (sendMessage, params) {
    await sbp('gi.actions/chatroom/rename', {
      ...omit(params, ['options', 'contractID', 'data', 'hooks']),
      contractID: params.data.chatRoomID,
      data: {
        name: params.data.name
      },
      hooks: {
        prepublish: params.hooks?.prepublish,
        postpublish: null
      }
    })

    return await sendMessage({
      ...omit(params, ['options', 'action', 'hooks']),
      hooks: {
        prepublish: null,
        postpublish: params.hooks?.postpublish
      }
    })
  }),
```

Notice:

1. You no longer need to pass in the `action` parameter to `sendMessage`.
2. You no longer need to wrap the inside of the function in `try/catch`.
3. You pass in the user-facing error as the 2nd parameter to `encryptedAction`

Reply:
Then, should I send this PR to `e2e-protocol` branch either?

Reply:
Yes, please branch off of `e2e-protocol` and send a PR to that branch.